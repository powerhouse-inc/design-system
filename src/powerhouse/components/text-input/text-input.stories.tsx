import type { Meta, StoryObj } from '@storybook/react';
import { TextInput } from './text-input';

const meta: Meta<typeof TextInput> = {
    title: 'Powerhouse/Components/TextInput',
    component: TextInput,
    argTypes: {
        startAdornment: { control: { type: 'text' } },
        endAdornment: { control: { type: 'text' } },
        inputProps: { control: { type: 'object' } },
        textFieldProps: { control: { type: 'object' } },
    },
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Primary: Story = {
    args: {
        startAdornment: <div className="p-2">Start</div>,
        endAdornment: <div className="p-2">End</div>,
        className: 'bg-white rounded-2xl p-2',
        textFieldProps: {
            autoFocus: true,
        },
    },
};
