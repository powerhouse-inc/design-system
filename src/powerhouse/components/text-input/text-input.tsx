import { mergeClassNameProps } from '@/powerhouse';
import React from 'react';
import {
    Input,
    InputProps,
    TextField,
    TextFieldProps,
} from 'react-aria-components';
import { twMerge } from 'tailwind-merge';

export interface TextInputProps
    extends Omit<React.HTMLAttributes<HTMLDivElement>, 'onChange'> {
    inputProps?: InputProps;
    endAdornment?: React.ReactNode;
    value?: TextFieldProps['value'];
    textFieldProps?: TextFieldProps;
    startAdornment?: React.ReactNode;
    onChange?: TextFieldProps['onChange'];
}

export const TextInput: React.FC<TextInputProps> = props => {
    const {
        value,
        onChange,
        className,
        endAdornment,
        startAdornment,
        inputProps = {},
        textFieldProps = {},
        ...containerProps
    } = props;

    return (
        <div
            className={twMerge('flex flex-row', className)}
            {...containerProps}
        >
            {startAdornment}
            <TextField
                value={value}
                onChange={onChange}
                {...mergeClassNameProps(textFieldProps, 'flex flex-1 min-w-0')}
            >
                <Input
                    {...mergeClassNameProps(
                        inputProps,
                        'flex flex-1 outline-none bg-inherit min-w-0',
                    )}
                />
            </TextField>
            {endAdornment}
        </div>
    );
};
