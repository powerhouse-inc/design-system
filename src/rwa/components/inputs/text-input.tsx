import { ComponentPropsWithRef, ForwardedRef, forwardRef, memo } from 'react';
import { twMerge } from 'tailwind-merge';

type Props = Omit<ComponentPropsWithRef<'input'>, 'type'> & {
    defaultValue?: string | number | null | undefined;
    errorMessage?: string;
    labelClassName?: string;
    inputClassName?: string;
    errorMessageClassName?: string;
};

export const _RWATableTextInput = forwardRef(function RWATableTextInput(
    props: Props,
    ref: ForwardedRef<HTMLInputElement>,
) {
    const invalid = props['aria-invalid'] === 'true';
    const {
        errorMessage,
        labelClassName,
        inputClassName,
        errorMessageClassName,
        ...inputProps
    } = props;

    return (
        <label className={labelClassName}>
            <input
                {...inputProps}
                type="text"
                ref={ref}
                className={twMerge(
                    'h-8 w-full rounded-md border border-transparent bg-gray-100 p-3 text-right placeholder:text-gray-500 disabled:bg-transparent disabled:p-0 disabled:text-left',
                    invalid &&
                        'border-red-900 outline-red-900 placeholder:text-red-800',
                    inputClassName,
                )}
            />
            {invalid && !!errorMessage && (
                <p
                    role="alert"
                    className={twMerge(
                        'text-sm text-red-900',
                        errorMessageClassName,
                    )}
                >
                    {errorMessage}
                </p>
            )}
        </label>
    );
});

export const RWATableTextInput = memo(_RWATableTextInput);
