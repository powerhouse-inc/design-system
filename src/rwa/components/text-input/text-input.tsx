import { TextInput, TextInputProps, mergeClassNameProps } from '@/powerhouse';
import React from 'react';

export interface RWATextInputProps extends TextInputProps {
    label?: string;
}

export const RWATextInput: React.FC<RWATextInputProps> = props => {
    const { label, ...textInputProps } = props;
    return (
        <div>
            {label && (
                <span className="pb-2 text-xs text-gray-600">{label}</span>
            )}
            <TextInput
                {...mergeClassNameProps(
                    textInputProps,
                    'h-[42px] rounded-xl bg-gray-100 px-3 text-xs',
                )}
            />
        </div>
    );
};
