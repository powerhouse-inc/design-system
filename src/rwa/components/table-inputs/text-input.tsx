import { useNumberFormat } from '@react-input/number-format';
import { Control, Controller, FieldValues, Path } from 'react-hook-form';
import { twMerge } from 'tailwind-merge';
import { RWATextInput, RWATextInputProps } from '../text-input';

export interface RWATableTextInputProps<ControlInputs extends FieldValues>
    extends RWATextInputProps {
    disabled?: boolean;
    name: Path<ControlInputs>;
    control: Control<ControlInputs>;
    type?: React.HTMLInputTypeAttribute | 'currency';
}

export function RWATableTextInput<ControlInputs extends FieldValues>(
    props: RWATableTextInputProps<ControlInputs>,
) {
    const {
        name,
        control,
        type = 'text',
        disabled = false,
        ...restProps
    } = props;

    const inputProps: RWATextInputProps['inputProps'] = { type };

    const currencyInputRef = useNumberFormat({
        locales: 'en',
        format: 'currency',
        currency: 'USD',
        maximumFractionDigits: 2,
    });

    if (type === 'currency') {
        inputProps.ref = currencyInputRef;
    }

    return (
        <Controller
            name={name}
            control={control}
            render={({ field: { onChange, onBlur, value } }) => (
                <RWATextInput
                    value={value}
                    onBlur={onBlur}
                    onChange={onChange}
                    textFieldProps={{ isDisabled: disabled }}
                    inputProps={{
                        className: 'text-right',
                        ...inputProps,
                    }}
                    className={twMerge(
                        'h-[32px] rounded-md',
                        disabled && 'bg-white p-0',
                    )}
                    {...restProps}
                />
            )}
        />
    );
}
