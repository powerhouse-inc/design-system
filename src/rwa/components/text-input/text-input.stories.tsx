import type { Meta, StoryObj } from '@storybook/react';
import { RWATextInput } from './text-input';

const meta: Meta<typeof RWATextInput> = {
    title: 'RWA/Components/RWATextInput',
    component: RWATextInput,
    argTypes: {
        inputProps: { control: { type: 'object' } },
        onChange: { control: { type: 'action' } },
        value: { control: { type: 'text' } },
        label: { control: { type: 'text' } },
    },
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Primary: Story = {
    args: {
        label: 'Label',
        inputProps: {
            placeholder: 'Placeholder',
        },
    },
};
