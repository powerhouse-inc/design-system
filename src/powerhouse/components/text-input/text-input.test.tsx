import { fireEvent, render, screen } from '@testing-library/react';
import { TextInput } from './text-input';

describe('Modal Component', () => {
    it('should match snapshot', () => {
        const { asFragment } = render(
            <TextInput
                startAdornment={<div>Start</div>}
                endAdornment={<div>End</div>}
                className="rounded-2xl bg-white p-2"
                textFieldProps={{ 'aria-label': 'test' }}
            />,
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it('should render start adornment', () => {
        render(
            <TextInput
                startAdornment={<div>Start</div>}
                endAdornment={<div>End</div>}
                className="rounded-2xl bg-white p-2"
                textFieldProps={{ 'aria-label': 'test' }}
            />,
        );

        expect(screen.getByText('Start')).toBeInTheDocument();
    });

    it('should render end adornment', () => {
        render(
            <TextInput
                startAdornment={<div>Start</div>}
                endAdornment={<div>End</div>}
                className="rounded-2xl bg-white p-2"
                textFieldProps={{ 'aria-label': 'test' }}
            />,
        );

        expect(screen.getByText('End')).toBeInTheDocument();
    });

    it('should call onChange when input value change', () => {
        const onChange = jest.fn();

        render(
            <TextInput
                startAdornment={<div>Start</div>}
                endAdornment={<div>End</div>}
                className="rounded-2xl bg-white p-2"
                textFieldProps={{ 'aria-label': 'test' }}
                onChange={onChange}
            />,
        );

        fireEvent.change(screen.getByRole('textbox'), {
            target: { value: 'test' },
        });

        expect(onChange).toHaveBeenCalled();
        expect(onChange).toHaveBeenCalledWith('test');
    });
});
