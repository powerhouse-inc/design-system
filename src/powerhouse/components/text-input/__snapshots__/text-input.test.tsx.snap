// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Modal Component > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-row rounded-2xl bg-white p-2"
  >
    <div>
      Start
    </div>
    <div
      class="flex flex-1 min-w-0"
      data-rac=""
    >
      <input
        aria-label="test"
        class="flex flex-1 outline-none bg-inherit min-w-0"
        data-rac=""
        id="react-aria-:r0:"
        title=""
        type="text"
        value=""
      />
    </div>
    <div>
      End
    </div>
  </div>
</DocumentFragment>
`;
